### PR TITLE
Support for a Collation Index

### PR DESCRIPTION
**Pull request by tcsiwula**
Added collation_index.rs and modified access to create sample header/body methods.

### CODE REVIEW

**Comment by timxor** [comment]
Let me know how this looks @jamesray1. I still need to add an insertion sort for the collation index, e.g. inserting a blob and serializing it into an existing collation. But if this data structure looks okay for now, then I will proceed :D also maybe what kind of tests you think would be good for this.

**Comment by jamesray1** [comment]
Just clarifying, you're aware that the Travis build is failing?

File: cli/src/lib.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ extern crate error_chain;
 extern crate log;
 
 pub mod error {
-    error_chain! { }
+//    error_chain! { }
```
Comment (jamesray1) [inline cli/src/lib.rs]: Why do you want to comment this out? We need to be integrating error_chain.

File: node/src/modules/collation/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# collation quick start
```
Comment (jamesray1) [inline node/src/modules/collation/README.md:1]: Good initiative on this README!

File: node/src/modules/collation/blob.rs [hunk 3]
```diff
@@ -505,4 +503,4 @@ pub fn clone_into_array<A, T>(slice: &[T]) -> A
         <A as AsMut<[T]>>::as_mut(&mut a).clone_from_slice(slice);
         a
     }
-*/
\ No newline at end of file
+*/
```
Comment (jamesray1) [inline node/src/modules/collation/blob.rs:506]: Were you using git diff to detect the whitespace? I've used that a few times, but should do it every time.

**Comment by jamesray1** [comment]
The build is failing on line 28 of src/lib.rs, due to an incorrect path, let me fix that before we merge this.

**Comment by jamesray1** [comment]
Working on this: https://github.com/Drops-of-Diamond/diamond_drops/pull/88/, but I've encountered another issue.

**Comment by jamesray1** [comment]
Fixed, restarted the build.

**Comment by jamesray1** [comment]
Ah I understand why just your build is failing, it's because you moved `create_sample_collation_body()` as into the `impl Body` (which is fine, thanks for that), however you didn't update the notation for calling `create_sample_collation_body()` in src/lib.rs (now on line 29, where the build is failing) to `body::Body::create_sample_collation_body()`.

**Comment by jamesray1** [comment]
Also, for future reference it's good practice to checkout to a new branch for features, and push changes from that new branch.

E.g. 
```
git checkout -b collation-index // make changes;
git add .; git commit -m "Add support for a collation index";
git push;`
```

Note that `git push` will prompt:

```
fatal: The current branch collation-index has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin collation-index
```

You can then copy (CTRL+SHIFT+C) and paste (CTRL+SHIFT+V) this line. This is more convenient than having to type it out.

File: cli/src/lib.rs (hunk 1 above)
Comment (timxor) [reply cli/src/lib.rs]: Error chain kept invoking linting errors on ```cargo test ```

File: node/src/modules/collation/blob.rs (hunk 3 above)
Comment (timxor) [reply node/src/modules/collation/blob.rs:506]: I have actually never used git diff, this may be related to the webstorm ide I was using and formating?

**Comment by timxor** [comment]
Whoops sorry about not checking them all. This is how you do that right? ``` cargo make test-all```

**Comment by timxor** [comment]
Will use branch names from now on, normally I do :D

**Comment by timxor** [comment]
Looks like it is still failing? Do you have travis set up to re-run the tests after updates on a pr?

**Comment by timxor** [comment]
Oh wait looks like it is good now: https://travis-ci.org/Drops-of-Diamond/diamond_drops/builds/386176027

File: cli/src/lib.rs (hunk 1 above)
Comment (jamesray1) [reply cli/src/lib.rs]: Can you copy and paste the error? Do you mean a warning?

File: node/src/modules/collation/blob.rs (hunk 3 above)
Comment (jamesray1) [reply node/src/modules/collation/blob.rs:506]: OK, I've been using VS Code, Jetbrains is another option (free, since we're open source).

**Comment by jamesray1** [comment]
This is failing but #90 is passing.

**Comment by jamesray1** [comment]
In VS Code you can use makefile.toml to run cargo make tasks from the GUI (rather than as commands in the terminal).

**Comment by jamesray1** [comment]
#90 supercedes this.

**Pull request closed without merging**